### PR TITLE
[framework] A few minor fix for dispatchable token 

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Description

Fixes a bunch of issues:
1. Updated move-unit-test to include full error msg on test failure so we can see the detailed error msg.
2. Fix error msg when the module that contains the hook function doesn't exist.
3. Fix error msg when invoking FS api on assets with hook registered.
4. Fix PFS mint to invoke withdraw_internal.
5. Fix transfer api in dispatchable api to check for the frozen bit.

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [x] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
Added tests for each affected cases.

## Key Areas to Review
N/A

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/dispatchable_fungible_asset.move [hunk 1]
```diff
@@ -72,6 +75,8 @@ module aptos_framework::dispatchable_fungible_asset {
                 features::dispatchable_fungible_asset_enabled(),
                 error::aborted(ENOT_ACTIVATED)
             );
+            fungible_asset::assert_not_frozen(store);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/dispatchable_fungible_asset.move:78]: Can you leave IF ELSE style comments on the corresponding function in fungible_asset and here to make sure any checks added in the future are present in both to avoid similar cases from happening?

File: third_party/move/move-vm/runtime/src/interpreter.rs [hunk 2]
```diff
@@ -663,30 +663,37 @@ impl Interpreter {
...
-                            .with_message(format!("Module {} failed to be charged", module_name))
-                    })?;
-                resolver
-                    .loader()
-                    .load_module(&module_name, data_store, module_store)
-                    .map_err(|_| {
-                        PartialVMError::new(StatusCode::FUNCTION_RESOLUTION_FAILURE)
-                            .with_message(format!("Module {} doesn't exist", module_name))
-                    })?;
+                if let Ok(true) = data_store.exists_module(&module_name) {
```
Comment (movekevin) [inline third_party/move/move-vm/runtime/src/interpreter.rs:666]: I can't tell what changed here. Is this just formatting?

File: aptos-move/framework/aptos-framework/sources/dispatchable_fungible_asset.move [hunk 3]
```diff
@@ -21,6 +21,7 @@ module aptos_framework::dispatchable_fungible_asset {
 
     use std::error;
     use std::features;
+    use std::signer;
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/dispatchable_fungible_asset.move:24]: nit: Sort this alphabetically

File: third_party/move/move-vm/runtime/src/interpreter.rs (hunk 2 above)
Comment (runtian-zhou) [reply third_party/move/move-vm/runtime/src/interpreter.rs:666]: I'm fixing the error msg when the module doesn't exist.

**Review by movekevin — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/dispatchable_fungible_asset.move (hunk 1 above)
Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/dispatchable_fungible_asset.move:83]: why not make it a friend function in `fungible_asset.move` called `sanity_check` or sth to reuse code and avoid omission.

File: aptos-move/framework/aptos-framework/sources/dispatchable_fungible_asset.move [hunk 4]
```diff
@@ -99,6 +109,12 @@ module aptos_framework::dispatchable_fungible_asset {
                 features::dispatchable_fungible_asset_enabled(),
                 error::aborted(ENOT_ACTIVATED)
             );
+            // ** IMPORTANT **
```
Comment (lightmark) [inline aptos-move/framework/aptos-framework/sources/dispatchable_fungible_asset.move:112]: same

**Review by lightmark — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move [hunk 5]
```diff
@@ -452,6 +453,11 @@ module aptos_framework::fungible_asset {
         store_exists(object::object_address(&store)) && borrow_store_resource(&store).frozen
```
Comment (lightmark) [inline aptos-move/framework/aptos-framework/sources/fungible_asset.move:453]: here

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move [hunk 6]
```diff
@@ -582,6 +588,11 @@ module aptos_framework::fungible_asset {
         store: Object<T>,
         amount: u64,
     ): FungibleAsset acquires FungibleStore, DispatchFunctionStore {
+        // ** IMPORTANT **
+        //
+        // For any changes in the access control logic here, make sure we duplicate them in
+        // dispatchable_fungible_asset::withdraw as that will be a separate entrypoint.
+        //
         assert!(object::owns(store, signer::address_of(owner)), error::permission_denied(ENOT_STORE_OWNER));
         assert!(store_exists(object::object_address(&store)), error::invalid_argument(ESTORE_IS_FROZEN));
```
Comment (lightmark) [inline aptos-move/framework/aptos-framework/sources/fungible_asset.move:597]: As I said in the landed PR, `borrow_store_resource` checks for you, please remove similar checks in this PR

**Review by lightmark — CHANGES_REQUESTED** [review]
Could we reuse the same piece of code in a function in fungible_asset.move for authorization and validation?
Also remove unnecessary existence check please. Thanks to @igor-aptos finding that

File: aptos-move/framework/aptos-framework/sources/primary_fungible_store.move [hunk 7]
```diff
@@ -203,7 +203,8 @@ module aptos_framework::primary_fungible_store {
     /// Mint to the primary store of `owner`.
     public fun mint(mint_ref: &MintRef, owner: address, amount: u64) acquires DeriveRefPod {
         let primary_store = ensure_primary_store_exists(owner, fungible_asset::mint_ref_metadata(mint_ref));
-        fungible_asset::mint_to(mint_ref, primary_store, amount);
+        let fa = fungible_asset::mint(mint_ref, amount);
+        fungible_asset::deposit_internal(primary_store, fa);
```
Comment (igor-aptos) [inline aptos-move/framework/aptos-framework/sources/primary_fungible_store.move:207]: why this change?

**Review by igor-aptos — APPROVED** [review]
move changes look good. 

can you get someone from VM team to look at error handling? maybe @georgemitenkov

File: third_party/move/move-vm/runtime/src/interpreter.rs [hunk 8]
```diff
@@ -663,30 +663,37 @@ impl Interpreter {
...
+                        .map_err(|err| {
+                            PartialVMError::new(err.major_status()).with_message(format!(
+                                "Module {} failed to be charged",
+                                module_name
+                            ))
+                        })?;
+                    resolver
+                        .loader()
+                        .load_module(&module_name, data_store, module_store)
+                        .map_err(|_| {
```
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/interpreter.rs:686]: Should we be propagating this error here?

File: aptos-move/framework/aptos-framework/sources/primary_fungible_store.move (hunk 7 above)
Comment (runtian-zhou) [reply aptos-move/framework/aptos-framework/sources/primary_fungible_store.move:207]: This is to make sure mint can work for dispatchable token using PFS api. otherwise mint would fail because fa::mint_to calls fa::deposit which would be rejected for dispatchable token.

File: third_party/move/move-vm/runtime/src/interpreter.rs (hunk 2 above)
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/interpreter.rs:666]: This call bypasses all caches, an issues a read to storage, loading all bytes. I wish we had on-demand loading so that we do not have to do this and just `load_module()?`. Cannot we do so already? Or the error message there does not work?

Comment (runtian-zhou) [reply third_party/move/move-vm/runtime/src/interpreter.rs:666]: The problem was that `check_dependencies_and_charge_gas` will throw really weird error when the module doesn't exist. Do you have a better idea how to do this properly? cc @vgao1996

Comment (georgemitenkov) [reply third_party/move/move-vm/runtime/src/interpreter.rs:666]: What is the error? I thought it is "Linker Error: Cannot find {:?} in data cache"?

Comment (runtian-zhou) [reply third_party/move/move-vm/runtime/src/interpreter.rs:666]: The error msg looks like: `LINKER_ERROR (code 1081) error with error message: "Module 000000000000000000000000000000000000000000000000000000000000face::deflation_token failed to be charged".` and it's a bit hard to read.

Comment (georgemitenkov) [reply third_party/move/move-vm/runtime/src/interpreter.rs:666]: Can we remove this error remapping all together? Seems like the error message should originate from the inside of `check_dependencies_and_charge_gas`?

Comment (runtian-zhou) [reply third_party/move/move-vm/runtime/src/interpreter.rs]: Updated the error msg! Now the error will look like this:
```
Linker Error: Module 000000000000000000000000000000000000000000000000000000000000face::deflation_token doesn't exist.Failed to charge transitive dependency for 000000000000000000000000000000000000000000000000000000000000face::deflation_token. Does this module exists?
```

**Review by georgemitenkov — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move [hunk 9]
```diff
@@ -667,8 +672,8 @@ module aptos_framework::fungible_asset {
 
     /// Mint the specified `amount` of the fungible asset to a destination store.
     public fun mint_to<T: key>(ref: &MintRef, store: Object<T>, amount: u64)
-    acquires FungibleStore, Supply, ConcurrentSupply, DispatchFunctionStore {
-        deposit(store, mint(ref, amount));
+    acquires FungibleStore, Supply, ConcurrentSupply {
+        deposit_internal(store, mint(ref, amount));
```
Comment (lightmark) [inline aptos-move/framework/aptos-framework/sources/fungible_asset.move:676]: Does this change the semantic? It seems this bypass the frozen flag?

File: aptos-move/framework/aptos-framework/sources/primary_fungible_store.move (hunk 7 above)
Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/primary_fungible_store.move:207]: I still don't get this... 
1. do you want to call fa::mint_to now since you changed the function.
2. why it would be rejected?

Comment (runtian-zhou) [reply aptos-move/framework/aptos-framework/sources/primary_fungible_store.move:207]: Lemme change it to fa::mint_to. This is the same issue.

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move (hunk 9 above)
Comment (runtian-zhou) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:676]: Good call. Lemme add the frozen flag again with tests!
